Add crystfel streams to `rs/io` module

marinegor opened the pull request:
This PR adds:
 - reading `CrystFEL's` stream files into `DataSet` format without any metadata
 - computing columns `ewald_offset`, `s1{x,y,z}` upon reading
 - accompanying tests
 - example of CrystFEL's stream to test on.

JBGreisman commented:
Overall, this looks mostly good to me. I'm gonna review and request a few minor revisions. 

High-level questions: 
- is a CrystFEL stream file only applicable to unmerged data? If so, we should explicitly set `DataSet.merged = False`
- as is, `DataSet.spacegroup` and `DataSet.cell` are left unset. Is the spacegroup or cell parameters accessible in the CrystFEL stream? It may be a good idea to parse them out if that's possible -- I'd imagine cell parameters may not be reasonable to get this way for a serial experiment, but perhaps they could be averaged in a meaningful way?
- It looks like `read_crystfel()` was given an optional `logfile=None` argument. I'm assuming this is a vestige from `read_precognition()`, but just wanted to check whether there are any additional arguments that this should take -- this could be a good place to add optional, user-specified spacegroup/cell parameters in case they aren't available in the stream itself.

JBGreisman commented:
I just noticed the test data is 98MB -- it looks like its a raw text with many "crystal" entries, that I'm assuming belong to a serial experiment. We try to keep this test data fairly minimal to what's needed to assess the core functionality. 

Do you mind reducing the data by only including something like 2 crystals? Whatever seems like a reasonable minimum number to test the parser.

JBGreisman commented on reciprocalspaceship/__init__.py:
```diff
@@ -18,21 +21,21 @@ def getVersionNumber():
 # Add support for MTZ data types:
 # see http://www.ccp4.ac.uk/html/f2mtz.html
 from .dtypes import (
```
Do you mind reverting this file back to original and only adding `read_crystfel` to the top-level namespace? I know the file isn't quite "conventionally" formatted, but I find the aligned comments most clear

marinegor replied:
Sure, do it. I just have `yapf` on by default, which messed already existing files.

JBGreisman commented on reciprocalspaceship/io/crystfel.py:
```diff
@@ -0,0 +1,190 @@
+import numpy as np
+
+
+def _parse_stream(filename: str) -> dict:
```
Nice parser -- clear to read. One revision -- please change the docstring to match that of `read_crystfel()`.

marinegor replied:
I didn't fully understand your remark regarding the docstring -- please check whether it's done as you intended.

JBGreisman commented on reciprocalspaceship/io/crystfel.py:
```diff
@@ -0,0 +1,190 @@
+
+                # calculate ewald offset and s1
+                hkl = np.array([h, k, l])
+                # print(A)
```
Please delete this line

JBGreisman commented on reciprocalspaceship/io/crystfel.py:
```diff
@@ -0,0 +1,190 @@
+    return answ_crystals
+
+
+def read_crystfel(streamfile, logfile=None) -> DataSet:
```
Please remove `logfile=None` from the call signature. It may be useful to add optional `spacegroup` and `cell`, but we can discuss elsewhere

marinegor replied:
Removed `logfile` and moved conversation about `spacegroup` and `cell` below

JBGreisman commented on reciprocalspaceship/io/crystfel.py:
```diff
@@ -0,0 +1,190 @@
+    ----------
+    streamfile : str of file
+        name of a .stream file or a file object
+    """
```
Please add:
```
Returns
--------
rs.DataSet
```

JBGreisman commented on reciprocalspaceship/io/crystfel.py:
```diff
@@ -0,0 +1,190 @@
+    mtztypes = ["H", "H", "H", "J", "Q", "I", "R", "R", "R", "R"]
+    dataset = DataSet()
+    for (k, v), mtztype in zip(df.items(), mtztypes):
+        dataset[k] = v
```
To my knowledge, these two lines can be simplified to:
`dataset[k] = v.astype(mtztype)`

marinegor replied:
I copied this line from `read_precognition.py` -- I believed you had some reason to do that :)

JBGreisman replied:
fair enough -- it certainly works as written, so we can just leave it. I cleaned up that section of `read_precognition()` last night, and we can always apply the same changes here.

JBGreisman commented on reciprocalspaceship/io/crystfel.py:
```diff
@@ -0,0 +1,190 @@
+                #  -63   41    9     -41.31      57.45     195.00     170.86  731.0 1350.4 p0
+                crystal_peak_number += 1
+                h, k, l, I, sigmaI, _, _, _, _, _ = [i for i in line.split()]
+                h, k, l = map(float, [h, k, l])
```
Should `h`, `k`, and `l` be cast to `int`? Or is it necessary to have floats for the subsequent matrix multply?

marinegor replied:
You're right, `int` should be enough. `numpy` casts them to float during the calculation.

JBGreisman commented on tests/io/test_crystfel.py:
```diff
@@ -0,0 +1,49 @@
+        self.assertTrue(len(hewl.index.unique()) < len(hewl.index))
+        self.assertTrue(
+            hewl['ewald_offset'].min() < 0 < hewl['ewald_offset'].max())
+        self.assertEqual(len(hewl.BATCH.unique()),
```
If we reduce the number of crystals in the test data we'll need to modify this assertion

JBGreisman reviewed (changes requested):
Thanks for the PR! On the whole it all looks great. All my comments were very minor. The only main things in my mind to change are:

1. Let's reduce the test file size
2. I think it would make sense to set the `spacegroup`, `cell`, and `merged` attributes, if at all possible

JBGreisman commented on tests/io/test_crystfel.py:
```diff
@@ -0,0 +1,49 @@
+        rs.io.read_crystfel(IOtest_mtz)
+
+
+class TestPrecognition(unittest.TestCase):
```
Rename to `TestCrystFEL` or something comparable

marinegor commented:
Hi @JBGreisman , thanks for your detailed reply. I resolved most of your comments, here is the detailed reply:

> is a CrystFEL stream file only applicable to unmerged data? If so, we should explicitly set DataSet.merged = False
Only unmerged, right. The flag is now set.

> Is the spacegroup or cell parameters accessible in the CrystFEL stream? 
During `stream` stage of processing, the symmetry is considered to be unknown. The only things that `stream` contains are cell parameters, lattice type and centering (you can see the notation yourself in the stream header, search for 'Begin unit cell` line. If you know how to specify them in the `Dataset.spacegroup`, I'm open for your suggestions.

> It may be a good idea to parse them out if that's possible -- I'd imagine cell parameters may not be reasonable to get this way for a serial experiment, but perhaps they could be averaged in a meaningful way?
Sure, I added parsing of cell parameters from the header and added respective test for equality (`np.allclose`-ity, to be precise).

> Do you mind reducing the data by only including something like 2 crystals? Whatever seems like a reasonable minimum number to test the parser.
I choose 3 -- like odd numbers. My only with was for the stream to have crystals with both positive and negative `ewald_offset` for one of the tests -- I overdid this, for sure.

JBGreisman commented:
good catch with the `BatchDtype`

JBGreisman reviewed (approved):
Looks good to me. I'll merge this in shortly.

JBGreisman commented:
Thanks @marinegor!
